### PR TITLE
add-cluster-cleanup-routine

### PR DESCRIPTION
**Pull request by satr**

### CODE REVIEW

**Review by nilsgstrabo — CHANGES_REQUESTED** [review]
Added some comments that we can discuss

File: radix-operator/main.go [hunk 1]
```diff
@@ -81,12 +83,15 @@ type App struct {
 	kubeUtil                 *kube.Kube
 	config                   *apiconfig.Config
 	kedaClient               kedav2.Interface
+	schedulers               []scheduler.TaskScheduler
 }
 
 func main() {
 	ctx, _ := signal.NotifyContext(context.Background(), syscall.SIGTERM, syscall.SIGINT)
+	var app *App
```
[the remark this answers is not in the document]
Comment (satr) [reply radix-operator/main.go]: Removed

File: radix-operator/main.go [hunk 2]
```diff
@@ -136,9 +141,22 @@ func initializeApp(ctx context.Context) (*App, error) {
 	if err != nil {
 		return nil, fmt.Errorf("failed to load ingress configuration: %w", err)
 	}
+	app.schedulers = app.createSchedulers(ctx)
```
[the remark this answers is not in the document]
Comment (satr) [reply radix-operator/main.go]: Moved

File: radix-operator/main.go [hunk 3]
```diff
@@ -161,6 +179,10 @@ func (a *App) Run(ctx context.Context) error {
 	g.Go(func() error { return alertController.Run(ctx, a.opts.alertControllerThreads) })
 	g.Go(func() error { return batchController.Run(ctx, 1) })
 	g.Go(func() error { return dnsAliasesController.Run(ctx, a.opts.environmentControllerThreads) })
+	g.Go(func() error {
```
[the remark this answers is not in the document]
Comment (satr) [reply radix-operator/main.go]: Fixed

File: radix-operator/scheduler/task_scheduler.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+func (e taskScheduler) Stop() {
+	e.cron.Stop()
+}
+
+func (e taskScheduler) Start() {
+	e.cron.Start()
+}
+
+// NewTaskScheduler Creates a new task scheduler
+func NewTaskScheduler(ctx context.Context, task tasks.Task, scheduleSpec string) (TaskScheduler, error) {
```
[the remark this answers is not in the document]
Comment (satr) [reply radix-operator/scheduler/task_scheduler.go:33]: Renamed

File: pkg/apis/utils/environments.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	"context"
+
+	"github.com/equinor/radix-common/utils"
+	"github.com/equinor/radix-operator/pkg/apis/radix/v1"
+	"github.com/equinor/radix-operator/pkg/client/clientset/versioned"
+	v2 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+)
+
+func UpdateRadixEnvironmentStatus(ctx context.Context, radixClient versioned.Interface, radixApplication *v1.RadixApplication, radixEnvironment *v1.RadixEnvironment, time v2.Time) error {
```
[the remark this answers is not in the document]
Comment (satr) [reply pkg/apis/utils/environments.go:13]: Moved

File: pkg/apis/applicationconfig/applicationconfig.go [hunk 6]
```diff
@@ -151,3 +156,38 @@ func (app *ApplicationConfig) OnSync(ctx context.Context) error {
 
 	return nil
 }
+
+func (app *ApplicationConfig) handleOrphanedEnvironments(ctx context.Context, syncTime metav1.Time) error {
```
[the remark this answers is not in the document]
Comment (satr) [reply pkg/apis/applicationconfig/applicationconfig.go:160]: Fixed

File: pkg/apis/radix/v1/radixenvironmenttypes.go [hunk 7]
```diff
@@ -36,4 +36,6 @@ type RadixEnvironmentSpec struct {
 type RadixEnvironmentStatus struct {
 	Reconciled meta.Time `json:"reconciled" yaml:"reconciled"`
 	Orphaned   bool      `json:"orphaned" yaml:"orphaned"`
+	// OrphanedTimestamp is a timestamp representing the server time when this RadixEnvironment was removed from the RadixApplication
+	OrphanedTimestamp string `json:"orphanedTimestamp" yaml:"orphanedTimestamp"`
```
[the remark this answers is not in the document]
Comment (satr) [reply pkg/apis/radix/v1/radixenvironmenttypes.go:40]: Changed

File: radix-operator/application/handler.go [hunk 8]
```diff
@@ -78,7 +79,7 @@ func (t *Handler) Sync(ctx context.Context, namespace, name string, eventRecorde
 	syncApplication := radixApplication.DeepCopy()
 	log.Ctx(ctx).Debug().Msgf("Sync application %s", syncApplication.Name)
 	applicationConfig := application.NewApplicationConfig(t.kubeclient, t.kubeutil, t.radixclient, radixRegistration, radixApplication, t.dnsConfig)
-	err = applicationConfig.OnSync(ctx)
+	err = applicationConfig.OnSync(ctx, metav1.NewTime(time.Now().UTC()))
```
[the remark this answers is not in the document]
Comment (satr) [reply radix-operator/application/handler.go:82]: Removed

File: pkg/apis/defaults/environment_variables.go [hunk 9]
```diff
@@ -229,4 +229,10 @@ const (
 	// RadixExternalRegistryDefaultAuthEnvironmentVariable Name of the secret containing default credentials for external container registries.
 	// Used when pulling images for components and jobs and for pulling images in Dockerfiles when building with buildah.
 	RadixExternalRegistryDefaultAuthEnvironmentVariable = "RADIX_EXTERNAL_REGISTRY_DEFAULT_AUTH_SECRET"
+
+	// RadixOrphanedEnvironmentsRetentionPeriodVariable The duration for which orphaned environments are retained
+	RadixOrphanedEnvironmentsRetentionPeriodVariable = "RADIXOPERATOR_ORPHANED_ENVIRONMENTS_RETENTION_PERIOD"
+
+	// RadixOrphanedEnvironmentsCleanupCronVariable The cron expression for cleaning up orphaned environments
+	RadixOrphanedEnvironmentsCleanupCronVariable = "RADIXOPERATOR_ORPHANED_ENVIRONMENTS_CLEANUP_CRONE"
```
[the remark this answers is not in the document]
Comment (satr) [reply pkg/apis/defaults/environment_variables.go:237]: Fixed

File: radix-operator/scheduler/task_scheduler.go [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+	taskLogger := internal.NewLogger(ctx)
+	c := cron.New(cron.WithLogger(taskLogger), cron.WithChain(cron.DelayIfStillRunning(taskLogger)))
+	if _, err := c.AddFunc(scheduleSpec, task.Run); err != nil {
+		return nil, err
+	}
+	log.Ctx(ctx).Info().Msgf("Created schedule %s for the task %s", scheduleSpec, task.String())
```
[the remark this answers is not in the document]
Comment (satr) [reply radix-operator/scheduler/task_scheduler.go:39]: Moved

File: radix-operator/main.go [hunk 11]
```diff
@@ -81,12 +83,15 @@ type App struct {
...
 	kedaClient               kedav2.Interface
+	schedulers               []scheduler.TaskScheduler
 }
 
 func main() {
 	ctx, _ := signal.NotifyContext(context.Background(), syscall.SIGTERM, syscall.SIGINT)
+	var app *App
 	go func() {
 		<-ctx.Done()
+		app.stopSchedulers()
```
[the remark this answers is not in the document]
Comment (satr) [reply radix-operator/main.go]: Fixed

File: charts/radix-operator/values.yaml [hunk 12]
```diff
@@ -125,6 +125,9 @@ ingress:
       annotations:
         nginx.org/lb-method: random
 
+task:
+  orphanedEnvironmentsRetentionPeriod: 24h
```
Comment (nilsgstrabo) [inline charts/radix-operator/values.yaml:129]: Should we use a higher default, e.g. 7 days (168 hours)?

**Review by nilsgstrabo — COMMENTED** [review]
Added a few comments that we can discuss

File: pkg/apis/kube/radix_environment.go [hunk 13]
```diff
@@ -31,6 +31,16 @@ func (kubeutil *Kube) GetEnvironment(ctx context.Context, name string) (*radixv1
 	return environment, nil
 }
 
+// ListEnvironmentsWithSelector Gets lists environments with label selector
+func (kubeutil *Kube) ListEnvironmentsWithSelector(ctx context.Context, labelSelectorString string) ([]*radixv1.RadixEnvironment, error) {
```
[the remark this answers is not in the document]
Comment (satr) [reply pkg/apis/kube/radix_environment.go:35]: Removed

File: pkg/apis/radix/v1/radixenvironmenttypes.go [hunk 14]
```diff
@@ -34,6 +34,8 @@ type RadixEnvironmentSpec struct {
 
 // RadixEnvironmentStatus is the status for an RE
 type RadixEnvironmentStatus struct {
-	Reconciled meta.Time `json:"reconciled" yaml:"reconciled"`
-	Orphaned   bool      `json:"orphaned" yaml:"orphaned"`
+	Reconciled metav1.Time `json:"reconciled" yaml:"reconciled"`
+	Orphaned   bool        `json:"orphaned" yaml:"orphaned"`
+	// OrphanedTimestamp is a timestamp representing the server time when this RadixEnvironment was removed from the RadixApplication
+	OrphanedTimestamp *metav1.Time `json:"orphanedTimestamp" yaml:"orphanedTimestamp"`
```
[the remark this answers is not in the document]
Comment (satr) [reply pkg/apis/radix/v1/radixenvironmenttypes.go:40]: Fixed

**Review by nilsgstrabo — APPROVED** [review]

**Pull request merged**